File: crave-on-api/app/Models/Order.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Order extends Model
{
    protected $fillable = [
        'user_id',
        'order_number',
        'status',
        'subtotal',
        'tax',
        'total_amount',
        'order_type',
        'delivery_address',
        'notes',
        'payment_status',
        'payment_method',
        'paid_at',
    ];

    protected function casts(): array
    {
        return [
            'subtotal'     => 'decimal:2',
            'tax'          => 'decimal:2',
            'total_amount' => 'decimal:2',
            'paid_at'      => 'datetime',
        ];
    }

    // ── Auto-generate Order Number on Create ──────────────

    protected static function booted(): void
    {
        static::creating(function (Order $order) {
            $order->order_number = static::generateOrderNumber();
        });
    }

    private static function generateOrderNumber(): string
    {
        $date   = now()->format('Ymd');
        $count  = static::whereDate('created_at', today())->count() + 1;
        return sprintf('ORD-%s-%04d', $date, $count);
        // e.g. ORD-20240520-0001
    }

    // ── Helpers ───────────────────────────────────────────

    // Only pending or confirmed orders can be cancelled
    public function isCancellable(): bool
    {
        return in_array($this->status, ['pending', 'confirmed']);
    }

    // ── Relationships ─────────────────────────────────────

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function items(): HasMany
    {
        return $this->hasMany(OrderItem::class);
    }
}

File: crave_on_mobile/lib/models/order.dart
class OrderItem {
  final int    id;
  final String productName;
  final double priceAtTime;
  final int    quantity;
  final double subtotal;

  OrderItem({
    required this.id,
    required this.productName,
    required this.priceAtTime,
    required this.quantity,
    required this.subtotal,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      id:           json['id'],
      productName:  json['product_name'],
      priceAtTime:  (json['price_at_time'] as num).toDouble(),
      quantity:     json['quantity'],
      subtotal:     (json['subtotal'] as num).toDouble(),
    );
  }
}

class Order {
  final int            id;
  final String         orderNumber;
  final String         status;
  final double         subtotal;
  final double         tax;
  final double         totalAmount;
  final String         orderType;
  final String?        deliveryAddress;
  final String?        notes;
  final String         paymentStatus;
  final String?        paymentMethod;
  final bool           isCancellable;
  final List<OrderItem> items;
  final String         createdAt;

  Order({
    required this.id,
    required this.orderNumber,
    required this.status,
    required this.subtotal,
    required this.tax,
    required this.totalAmount,
    required this.orderType,
    this.deliveryAddress,
    this.notes,
    required this.paymentStatus,
    this.paymentMethod,
    required this.isCancellable,
    required this.items,
    required this.createdAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id:              json['id'],
      orderNumber:     json['order_number'],
      status:          json['status'],
      subtotal:        (json['subtotal'] as num).toDouble(),
      tax:             (json['tax'] as num).toDouble(),
      totalAmount:     (json['total_amount'] as num).toDouble(),
      orderType:       json['order_type'],
      deliveryAddress: json['delivery_address'],
      notes:           json['notes'],
      paymentStatus:   json['payment_status'],
      paymentMethod:   json['payment_method'],
      isCancellable:   json['is_cancellable'] ?? false,
      items:           (json['items'] as List? ?? [])
          .map((i) => OrderItem.fromJson(i))
          .toList(),
      createdAt:       json['created_at'],
    );
  }
}